File: Leetcode/Java/687. Longest Univalue Path/solution.java
// https://leetcode.com/problems/longest-univalue-path/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int longest = 0;

    public int longestUnivaluePath(TreeNode root) {
        this.dfs(-1, root);
        return this.longest;
    }

    private int dfs(int val, TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = dfs(node.val, node.left);
        int right = dfs(node.val, node.right);
        this.longest = Math.max(this.longest, left + right);

        if (val != node.val) {
            return 0;
        }
        return Math.max(left, right) + 1;
    }
}

File: Leetcode/Typescript/687. Longest Univalue Path/solution.ts
// https://leetcode.com/problems/longest-univalue-path/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function longestUnivaluePath(root: TreeNode | null): number {
  let length = 0;

  function dfs(root: TreeNode | null, parentNum: number): number {
    if (!root) {
      return 0;
    }
    const left = dfs(root.left, root.val),
      right = dfs(root.right, root.val);
    length = Math.max(length, left + right);

    if (root.val !== parentNum) {
      return 0;
    }
    return Math.max(left, right) + 1;
  }

  dfs(root, -1);

  return length;
}
